Add basic WDL 1.0 support

w-gao opened the pull request:
We should probably add tests for this, but we can run 1.0 WDL files now with a few exceptions:

- [Import statements](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#import-statements) are not supported
- [Object literals](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#object-literals) are not supported
- [Structs](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#custom--types) are not supported

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * Add basic WDL 1.0 support.

adamnovak commented on src/toil/wdl/versions/v1.py:
```diff
@@ -33,4 +48,736 @@ def analyze(self):
+        """
+        Expression placeholder options.
+
+        Matches three of the following:
```
Do you mean "all three of the following"?

adamnovak commented on src/toil/wdl/versions/v1.py:
```diff
@@ -33,4 +48,736 @@ def analyze(self):
+            return self.visitArray_literal(expr)
+        elif is_context(expr, 'Pair_literalContext'):
+            return self.visitPair_literal(expr)
+        # elif is_context(expr, 'Map_literalContext'):
+        #     raise NotImplementedError
+        # elif is_context(expr, 'Object_literalContext'):
+        #     raise NotImplementedError
```
Did you mean to leave these in here? I might replace them with a `# TODO: implement XXX and YYY`.

adamnovak commented on src/toil/wdl/versions/v1.py:
```diff
@@ -33,4 +48,736 @@ def analyze(self):
+            return self.visitUnarysigned(expr)
+        elif is_context(expr, 'PrimitivesContext'):
+            return self.visitPrimitives(expr)
+        # elif is_context(expr, 'Left_nameContext'):
+        #     raise NotImplementedError
```
Similarly, is this supposed to be like this?

adamnovak commented on src/toil/wdl/versions/v1.py:
```diff
@@ -33,4 +48,736 @@ def analyze(self):
+
+        return f'({expr_core}_{identifier})'
+
+    def visitNegate(self, ctx):
+        """
+        Pattern: NOT expr
+        """
+        return f'(not {self.visitExpr(ctx.expr())})'
+
+    def visitUnarysigned(self, ctx):
+        """
+        Pattern: (PLUS | MINUS) expr
+        """
+        plus: bool = ctx.PLUS() is not None
+        expr = self.visitExpr(ctx.expr())
+
+        if plus:
+            return f'(+{expr})'
+        return f'(-{expr})'
```
How come these are down here and not with the other operator code for e.g. + and -?

w-gao replied:
This is defined like that in the [grammar file](https://github.com/openwdl/wdl/blob/main/versions/1.0/parsers/antlr4/WdlV1Parser.g4#L132). I suspect that it has something to do with the order of operations, but I could be wrong.

adamnovak reviewed (changes requested):
This looks pretty good! And like it must have been a lot of work!

There are a few comments I want to be different, and some commented-out code that I think shouldn't be there, but overall this is great!

w-gao commented on src/toil/wdl/versions/v1.py:
```diff
@@ -33,4 +48,736 @@ def analyze(self):
+        """
+        element = ctx.children[0]
+        # task
+        if is_context(element, 'TaskContext'):
```
I was thinking about subclassing this class for the development version, so we can reuse a lot of the code. 

Using `isinstance` would prevent us from doing that, because antlr4 will generate different parsers and classes (e.g.: `WdlV1Parser.TaskContext` and `WdlParser.TaskContext`) based on the grammar file. 

Please let me know if there is a better way!

adamnovak replied:
I think that makes sense. We might want to explain it in a docstring or comment, though, instead of just in a discussion deep in the Github comment history.

w-gao commented on src/toil/wdl/versions/v1.py:
```diff
@@ -33,4 +48,736 @@ def analyze(self):
+                wf[wf_key] = contents
+            # parameter_meta and meta
+            elif is_context(section, ('Parameter_meta_elementContext', 'Meta_elementContext')):
+                pass
```
I don't think we need to do anything for these for now, so I'll go ahead and add a comment.

w-gao commented on src/toil/wdl/versions/v1.py:
```diff
@@ -33,4 +48,736 @@ def analyze(self):
+        """
+        return [self.visitAny_decls(decl) for decl in ctx.any_decls()]
+
+    def visitTask_output(self, ctx):
```
These are inherited methods from `WdlV1ParserVisitor`. Antlr4 generates the visitor methods by prepending `visit` + the capitalized token names. Unfortunately the [WDL grammar file](https://github.com/openwdl/wdl/blob/main/versions/1.0/parsers/antlr4/WdlV1Parser.g4#L214) uses snake case for the token names, so the generated methods end up looking like that. 

We actually don't need to subclass the visitor class if we don't rely on any of the parent methods, which is simple to do. I think that way we can change the naming scheme to snake case.

adamnovak replied:
We can consider doing that, but if there's a reason for these names I think they are fine to keep.

w-gao commented:
@adamnovak I think I addressed all of your comments. I left the ones that I wasn’t sure if I fully addressed as unresolved, and I’d be happy to revise.

I also added a small test, so it would be greatly appreciated if you could review that as well!

adamnovak reviewed (approved):
I still wish the string-type-checking system had a better explanation in the code about why it is the way it is. But I now know why it is the way it is (which is so that it's portable across multiple similar auto-generated parsers), so I'm happy with it.

The new tests also look good.

DailyDreaming reviewed (approved):
This looks good to me and I am merging.
